packages/executor/src/index.ts

```diff
@@ -37,89 +44,138 @@ export class ChugSplashExecutor extends BaseServiceV2<Options, Metrics, State> {
-      this.state.registry.filters.EventAnnounced('ChugSplashBundleApproved')
+      this.state.registry.filters.EventAnnounced('ChugSplashBundleApproved'),
+      this.state.lastBlockNumber + 1
     )
 
+    // If none found, return
+    if (approvalAnnouncementEvents.length === 0) {
+      this.logger.info('no events found')
+      return
+    }
+
+    this.logger.info(`${approvalAnnouncementEvents.length} events found`)
+
+    // store last block number
+    this.state.lastBlockNumber = approvalAnnouncementEvents.at(-1).blockNumber
+
+    // execute all approved bundles
     for (const approvalAnnouncementEvent of approvalAnnouncementEvents) {
+      // fetch manager for relevant project
       const signer = this.state.wallet
       const manager = new ethers.Contract(
         approvalAnnouncementEvent.args.manager,
         ChugSplashManagerABI,
         signer
       )
 
+      // get active bundle id for this project
       const activeBundleId = await manager.activeBundleId()
       if (activeBundleId === ethers.constants.HashZero) {
-        console.log('no active bundle')
+        this.logger.error(`Error: No active bundle id found in manager`)
         continue
       }
 
+      // fetch bundle state
       const bundleState: ChugSplashBundleState = await manager.bundles(
         activeBundleId
       )
 
-      // TODO: Add this to the ChugSplashManager contract
-      const selectedExecutor = await manager.getSelectedExecutor(activeBundleId)
-      if (selectedExecutor !== ethers.constants.AddressZero) {
-        // Someone else has been selected to execute the upgrade, so we can skip it.
-        continue
-      }
-
+      // get proposal event and compile
       const proposalEvents = await manager.queryFilter(
         manager.filters.ChugSplashBundleProposed(activeBundleId)
       )
-
-      if (proposalEvents.length !== 1) {
-        // TODO: throw an error here or skip
-      }
-
       const proposalEvent = proposalEvents[0]
       const { bundle, canonicalConfig } = await compileRemoteBundle(
         hre,
         proposalEvent.args.configUri
       )
+
+      // ensure compiled bundle matches proposed bundle
       if (bundle.root !== proposalEvent.args.bundleRoot) {
-        // TODO: throw an error here or skip
+        // log error and continue
+        this.logger.error(
+          'Error: Compiled bundle root does not match proposal event bundle root',
+          canonicalConfig.options
+        )
+        continue
+      }
+
+      // execute bundle
+      try {
+        await hre.run('chugsplash-execute', {
+          chugSplashManager: manager,
+          bundleState,
+          bundle,
+          parsedConfig: canonicalConfig,
+          deployer: signer,
+          hide: false,
+        })
+      } catch (e) {
+        // log error and continue
+        this.logger.error('Error: execution error', e, canonicalConfig.options)
+        continue
+      } finally {
+        this.logger.info('Successfully executed')
```

Comment:
I think this should go on line 147 since code inside of a 'finally' statement will execute when an exception is thrown (i.e. when execution fails)